File: src/beacon_detector/beacon_detector/PositionFinder.py
import numpy as np
from .math_lidar import *


class PositionFinder:

    def __init__(
        self, fixed_beacons, boundaries, precision=0.1, init_position=None
    ) -> None:
        self.previous_robot = None
        self.current_robot = None
        self.true_valor = False
        self.initialisation = False
        self.registre_init = []
        self.tour_repr = 0
        self.tour_reinit = 0
        self.tour_test = 0
        self.precision = precision
        self.BaliseA = fixed_beacons[0]
        self.BaliseB = fixed_beacons[1]
        self.BaliseC = fixed_beacons[2]
        self.BaliseD = fixed_beacons[3]
        self.boundaries = boundaries
        self.fixed_beacons = fixed_beacons
        if init_position is not None:
            self.previous_robot = RobotDatas()
            self.previous_robot.position = Point(
                x=init_position[0], y=init_position[1], z=np.pi * init_position[2] / 180
            )
            self.previous_robot.balise_A = chgt_base_plateau_to_robot(
                self.fixed_beacons[0], self.previous_robot.position
            )
            self.previous_robot.balise_B = chgt_base_plateau_to_robot(
                self.fixed_beacons[1], self.previous_robot.position
            )
            self.previous_robot.balise_C = chgt_base_plateau_to_robot(
                self.fixed_beacons[2], self.previous_robot.position
            )
            self.previous_robot.balise_D = chgt_base_plateau_to_robot(
                self.fixed_beacons[3], self.previous_robot.position
            )
            self.initialisation = True

    # 1
    def _update_data(self) -> None:
        self.tour_repr += 1
        if not self.true_valor and self.initialisation:
            self.tour_test += 1
        elif self.true_valor:
            self.tour_test = 0
        if self.tour_test >= 30:
            self.tour_test = 0
            self.initialisation = False
        if not self.initialisation:
            self.tour_reinit += 1

    # 3
    def _compare_previous_found_positions(self, potential_robot_datas) -> RobotDatas:
        robot_datas = potential_robot_datas.copy()
        dst_min = robot_datas[0].erreur
        best_match = robot_datas[0]
        for robot_data in robot_datas[1:]:
            dst_test = robot_data.erreur
            if dst_test < dst_min:
                dst_min = dst_test
                best_match = robot_data
        if dt(self.previous_robot.position, best_match.position) < self.precision:
            self.current_robot = best_match
            self.true_valor = True
        else:
            self.true_valor = False
            self.current_robot = self.previous_robot

    # 4
    def _recreate_balise(self) -> None:
        if self.current_robot.position.z is not None:
            if self.current_robot.balise_A is None:
                self.current_robot.balise_A = chgt_base_plateau_to_robot(
                    self.BaliseA, self.current_robot.position
                )
            if self.current_robot.balise_B is None:
                self.current_robot.balise_B = chgt_base_plateau_to_robot(
                    self.BaliseB, self.current_robot.position
                )
            if self.current_robot.balise_C is None:
                self.current_robot.balise_C = chgt_base_plateau_to_robot(
                    self.BaliseC, self.current_robot.position
                )
            if self.current_robot.balise_D is None:
                self.current_robot.balise_D = chgt_base_plateau_to_robot(
                    self.BaliseD, self.current_robot.position
                )

    # 5
    def _find_robots_on_plateau(self, obstacles) -> None:
        self.current_robot.other_robots = []
        cos_theta = np.cos(self.current_robot.position.z)
        sin_theta = np.sin(self.current_robot.position.z)
        for obstacle in obstacles:
            x, y = obstacle.x, obstacle.y
            x1, y1 = (cos_theta * x - sin_theta * y + self.current_robot.position.x), (
                cos_theta * y + sin_theta * x + self.current_robot.position.y
            )
            if (
                (x1 > self.boundaries[0])
                and (x1 < self.boundaries[1])
                and (y1 > self.boundaries[2])
                and (y1 < self.boundaries[3])
            ):
                if (
                    (
                        not approx(
                            dt(obstacle, self.current_robot.balise_A), precision=0.01
                        )
                    )
                    and (
                        not approx(
                            dt(obstacle, self.current_robot.balise_B), precision=0.01
                        )
                    )
                    and (
                        not approx(
                            dt(obstacle, self.current_robot.balise_C), precision=0.01
                        )
                    )
                    and (
                        not approx(
                            dt(obstacle, self.current_robot.balise_D), precision=0.01
                        )
                    )
                ):
                    positionement = np.array([x1, y1, 1])
                    self.current_robot.other_robots.append(positionement)

    # 2
    def _init_finding(self, potential_robot_datas) -> None:
        if len(self.registre_init) == 0:
            liste_tool_globale = []
            liste_balise = potential_robot_datas.copy()
            while len(liste_balise) > 0:
                ver = False
                bal = liste_balise.pop()
                for element in liste_tool_globale:
                    if not ver:
                        if (
                            abs(element[0].position.x - bal.position.x) < self.precision
                            and abs(element[0].position.y - bal.position.y)
                            < self.precision
                        ):
                            ver = True
                            element[1] += 1
                            element[0].position.x = (
                                element[0].position.x * (element[1] - 1)
                                + bal.position.x
                            ) / element[1]
                if not ver:
                    liste_tool_globale.append([bal, 1])

            self.registre_init.append(liste_tool_globale)
        else:
            liste_tool_globale = []
            liste_balise = potential_robot_datas.copy()
            liste_prev = self.registre_init[-1].copy()
            liste_rm = liste_prev.copy()
            while len(liste_balise) > 0:
                ver = False
                bal = liste_balise.pop()
                for element in liste_tool_globale:
                    if not ver:
                        if (
                            abs(element[0].position.x - bal.position.x) < self.precision
                            and abs(element[0].position.y - bal.position.y)
                            < self.precision
                        ):
                            ver = True
                            element[1] += 1
                            element[0].position.x = (
                                element[0].position.x * (element[1] - 1)
                                + bal.position.x
                            ) / element[1]
                if not ver:
                    for element in liste_prev:
                        if not ver:
                            if (
                                abs(element[0].position.x - bal.position.x)
                                < self.precision
                                and abs(element[0].position.y - bal.position.y)
                                < self.precision
                            ):
                                ver = True
                                liste_tool_globale.append([bal, 1 + element[1]])
                                liste_rm.remove(element)
                if not ver:
                    liste_tool_globale.append([bal, 1])
                liste_prev = liste_rm.copy()
            for element in liste_prev:
                liste_tool_globale.append(element)
            self.registre_init.append(liste_tool_globale)
        if self.tour_reinit >= 4:
            if len(self.registre_init) > 0:
                if len(self.registre_init[-1]) > 0:
                    donnees = self.registre_init[-1][0][0]
                    max = self.registre_init[-1][0][1]
                    for balise in self.registre_init[-1][1:]:
                        if balise[1] > max:
                            max = balise[1]
                            donnees = balise[0]
                    donnees.position.z = find_angle_31a(
                        donnees,
                        donnees.position.x,
                        donnees.position.y,
                        self.fixed_beacons,
                    )
                    self.current_robot = donnees
                    self._recreate_balise()
                    self.true_valor = True
                    self.initialisation = True
            self.tour_reinit = 0
            self.tour_test = 0
            self.registre_init = []

    # Used in beacon_detector_node.py
    def donnees_finales(
        self, nb_potential_beacons, potential_beacons, obstacles
    ) -> None:
        self._update_data()
        if nb_potential_beacons > 0:
            potential_robot_datas, _, _, _ = compute_potential_positions(
                potential_beacons,
                self.fixed_beacons,
                nb_potential_beacons,
                self.boundaries,
            )

            if len(potential_robot_datas) > 0:
                if not self.initialisation:
                    self._init_finding(potential_robot_datas)
                else:
                    self._compare_previous_found_positions(potential_robot_datas)
                if self.initialisation and self.true_valor:
                    self._recreate_balise()
                    self._find_robots_on_plateau(obstacles)
                    return self.current_robot
        self.true_valor = False
        return None

File: src/localization/src/PositionFinder.cpp
#include "localization/PositionFinder.hpp"

PositionFinder::PositionFinder(
    const std::array<Eigen::Vector2d, 4>& fixed_beacons,
    const std::array<double, 4>& boundaries,
    double precision,
    std::optional<std::array<double, 3>> init_position)
    : fixed_beacons_(fixed_beacons), // Real Beacons existing
      boundaries_(boundaries), // Boundaries of the board
      precision_(precision), // Precision required
      true_valor_(false), // Has the valor been found?
      initialisation_(false), // Is the node init?
      tour_repr_(0), // Useless
      tour_reinit_(0), // Init laps
      tour_test_(0) // Test false value incr
{
    if (init_position.has_value())
    {
        Eigen::Vector3d vec_init(init_position.value()[0], init_position.value()[1], init_position.value()[2]);
        previous_robot = vec_init;
        initialisation_ = true;
    }
}

void PositionFinder::update_data()
{
    tour_repr_++;
    // If the robot is initialized but not located well, we incr the test
    if (!true_valor_ && initialisation_)
    {
        tour_test_++;
    }
    // If the position seems good, we reset the test laps
    else if (true_valor_)
    {
        tour_test_ = 0;
    }
    // We check if the robot is lost for too long
    if (tour_test_ >= 15)
    {
        tour_test_ = 0;
        initialisation_ = false;
    }

    // If the robot is not initialized, we incr the laps of init
    if (!initialisation_)
    {
        tour_reinit_++;
    }
}

void PositionFinder::compare_previous_found_positions(
    const std::vector<std::pair<Eigen::Vector3d, double>>& potential_positions)
{
    Eigen::Vector3d best_match = potential_positions[0].first;
    double min_err = potential_positions[0].second;

    for (std::size_t i = 1; i < potential_positions.size(); i++)
    {
        double err = potential_positions[i].second;
        if (err < min_err)
        {
            min_err = err;
            best_match = potential_positions[i].first;
        }
    }
    Eigen::Vector2d p1(best_match.x(), best_match.y());
    Eigen::Vector2d p2(previous_robot->x(), previous_robot->y());
    if (dt(p1, p2) < precision_)
    {
        current_robot = best_match;
        true_valor_ = true;
    }
    else
    {
        true_valor_ = false;
        current_robot = previous_robot;
    }
}

std::array<Eigen::Vector2d, 4> PositionFinder::recreate_beacons(const Eigen::Vector3d& robot_frame)
{
    std::array<Eigen::Vector2d, 4> robot_beacons;
    for (int i = 0; i < 4; ++i)
    {
        robot_beacons[i] = chgt_base_plateau_to_robot(fixed_beacons_[i], robot_frame).head<2>();
    }
    return robot_beacons;
}

std::vector<Eigen::Vector2d> PositionFinder::find_robots_on_plateau(const std::vector<Eigen::Vector2d>& new_objects_detected)
{
    std::vector<Eigen::Vector2d> other_robots;
    double cos_theta = std::cos(current_robot->z());
    double sin_theta = std::sin(current_robot->z());
    Eigen::Matrix3d OtoR;
    OtoR << cos_theta, -sin_theta, current_robot->x(),
        sin_theta, cos_theta, current_robot->y(),
        0, 0, 1;

    for (const auto& obstacle : new_objects_detected)
    {
        Eigen::Vector3d oP = OtoR * Eigen::Vector3d(obstacle.x(), obstacle.y(), 1);
        if (oP.x() > boundaries_[0] && oP.x() < boundaries_[1] &&
            oP.y() > boundaries_[2] && oP.y() < boundaries_[3] &&
            !(abs(oP.x() - current_robot->x()) < 0.2 &&
            abs(oP.y() - current_robot->y()) < 0.2 ))
        {
            Eigen::Vector2d other(oP.x(), oP.y());
            other_robots.push_back(other);
        }
    }
    return other_robots;
}

void PositionFinder::init_finding(
    const std::vector<std::pair<Eigen::Vector3d, double>>& potential_positions)
{
    for (auto pos : potential_positions)
    {
        Eigen::Vector2d p1(pos.first.x(), pos.first.y());
        bool found = false;
        std::size_t i = 0;
        while (!found && i < registre_init_.size())
        {
            Eigen::Vector2d p2(pos.first.x(), pos.first.y());
            if (dt(p1, p2) < precision_)
            {
                found = true;
                registre_init_[i].second += 1;
                registre_init_[i].first = pos.first;
            }
        }
        if (!found)
        {
            std::pair<Eigen::Vector3d, int> reg;
            reg.first = pos.first;
            reg.second = 1;
            registre_init_.push_back(reg);
        }
    }
    if (tour_reinit_ >= 4 && !registre_init_.empty())
    {
        Eigen::Vector3d best_match = registre_init_[0].first;
        int max_count = registre_init_[0].second;
        for (std::size_t i = 1; i < registre_init_.size(); i++)
        {
            if (registre_init_[i].second > max_count)
            {
                max_count = registre_init_[i].second;
                best_match = registre_init_[i].first;
            }
        }
        current_robot = best_match;
        true_valor_ = true;
        initialisation_ = true;
        tour_reinit_ = 0;
        tour_test_ = 0;
        registre_init_.clear();
    }
    else if (tour_reinit_ >= 4)
    {
        current_robot = std::nullopt;
        true_valor_ = false;
        initialisation_ = false;
        tour_reinit_ = 0;
        tour_test_ = 0;
        registre_init_.clear();
    }
}

std::optional<Eigen::Vector3d> PositionFinder::search_pos(
    int nb_potential_beacons,
    const std::vector<std::array<std::optional<Eigen::Vector2d>, 4>>& potential_beacons,
    const std::vector<Eigen::Vector2d>& new_objects_detected)
{
    update_data();
    std::vector<std::pair<Eigen::Vector3d, double>> potential_positions = find_position(potential_beacons, fixed_beacons_, nb_potential_beacons, boundaries_);
    if (!potential_positions.empty())
    {
        if (!initialisation_)
        {
            init_finding(potential_positions);
        }
        else
        {
            compare_previous_found_positions(potential_positions);
        }

        if (initialisation_ && true_valor_)
        {
            find_robots_on_plateau(new_objects_detected);
            return current_robot;
        }
        else
        {
            return std::nullopt;
        }
    }
    true_valor_ = false;
    return std::nullopt;
}
